### ISSUE TITLE
User import gets exponentially slow

### ISSUE THREAD

**Issue by SebastEnn** [keycloak/keycloak#34364]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

import-export

### Describe the bug

I have an Keycloak instance with 10k users. When doing an upgrade to a new version, we export all users in multiple files with 50 users per file (default from Keycloak) and then import the new realm config and all those user files.

During import I noticed that the time it takes to import a single users file is about 1 second on the first file and with each new files it increases exponentially up to 20 seconds per file.

While researching this bug i found the issue #20703. 
It was resolved by adding this note to the [Importing and Exporting Realms](https://www.keycloak.org/server/importExport) guide:
> Exporting and importing into single files can produce large files, so if your database contains more than 500 users, export to a directory and not a single file. 
> Using a directory performs better as the directory provider **uses a separate transaction** for each "page" (a file of users).

Apparently this doesn't apply to the import. My guess is: All user files are being imported in **a single transaction**, which get's bigger and bigger with each batch of 50 users, resulting in exponentially longer import times with each file.

### Version

26.0.2

### Regression

- [ ] The issue is a regression

### Expected behavior

The time it takes to import a single users file stays the same.

### Actual behavior

The time it takes to import a single users file increases exponentially with each file imported.

### How to Reproduce?

- Get yourself an Keycloak instance with a few thousand users
- (CLI) Export the realm with these flags: --users different_files --users-per-file 50
- (CLI) Import the realm and all user files from the directory to which they were exported
- Notice the time it takes for each file to be import increasing with each import

**Comment by SebastEnn** [comment]
I found the (probably) responsible line of code.

https://github.com/keycloak/keycloak/blob/dbc2ab3cd060ee34a947722ed9c119fdf24c8762/model/storage-services/src/main/java/org/keycloak/exportimport/dir/DirImportProvider.java#L157

**Comment by pedroigor** [comment]
I agree with your guess. We probably want to import users in batches and separate transactions. We have a similar mechanism, so solve the same problem, when provisioning users using the `keycloak-benchmark` tool.

**Comment by pedroigor** [comment]
Marking as `low` priority because you can also upgrade to a new version by backing up the database. This is the recommendation.

However, it makes sense to improve the import.

**Comment by upadhyayniti** [comment]
Please note that export is also taking longer than expected. When fixing and testing import, please include export as well.

**Comment by SebastEnn** [comment]
@Huluti if you really need a quick fix without changing your Keycloak setup:
You could start up a temporary Keycloak instance connected to your db and then run a script in your preferred language, which goes through every user file asynchronously and imports it on the [PartialImport](https://www.keycloak.org/docs-api/latest/rest-api/index.html#_post_adminrealmsrealmpartialimport) endpoint.

That's what I did for now, but this should only be a temporary fix.

**Comment by Huluti** [comment]
Thanks @SebastEnn, I'll consider this option. It may be the best way for now to unblock my migration.

**Comment by modax** [comment]
@shawkins Well, at 26.1.0 state, import gets exponentially slower with each user file (or rather user entity) because transaction builds up or whatever JPA/Hibernate is doing there (I'm not an expert of JPA/Hibernate). With that patch, each import of each user file is as fast as the first one. IIRC, 10k+ users took 1-2 minutes for me (with 50 users per file) while I have never waited for import to complete on unpatched Keycloak (I think it would have taken many hours).

Yeah, I realize that my workaround is not a real solution. But I needed a quick workaround because all my migration strategy was on based on export/import.

**Comment by shawkins** [comment]
> Please note that export is also taking longer than expected. When fixing and testing import, please include export as well.

@upadhyayniti can you please create a separate issue for this? What performance are you seeing and for how many users?
